src/Compilers/Core/Portable/SourceGeneration/AdditionalSourcesCollection.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    // PROTOTYPE: should this implement ICollection?
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "In Progress")]
+    public class AdditionalSourcesCollection
+    {
+        private readonly PooledDictionary<string, SourceText> _sourcesAdded;
+
+        internal AdditionalSourcesCollection()
+        {
+            _sourcesAdded = PooledDictionary<string, SourceText>.GetInstance();
+        }
+
+        internal AdditionalSourcesCollection(ImmutableArray<GeneratedSourceText> sources)
+            : this()
+        {
+            foreach (var source in sources)
+            {
+                _sourcesAdded.Add(source.Hint, source.Text);
+            }
+        }
+
+        public void Add(string hintName, SourceText source)
+        {
+            if (Contains(hintName))
+            {
+                throw new ArgumentException($"Parameter {nameof(hintName)} must be a unique value.", nameof(hintName));
+            }
+            _sourcesAdded.Add(hintName, source);
+        }
+
+        public void RemoveSource(string hintName)
```

Comment:
Agreed. Currently we only ever new one up per-generator/update, so it's only ever accessed by a single generator in a single invocation at a time.

---
In reply to: [382209903](https://github.com/dotnet/roslyn/pull/41789#discussion_r382209903) [](ancestors = 382209903)

Reply:
@chsienki So what's the idea behind this method? A generator generated something, and then had second thoughts?

Reply:
During an edit the compilation might change (for instance a type is removed, or an additional file deleted) that would cause a previously generated source file to no longer be produced.

---
In reply to: [388636623](https://github.com/dotnet/roslyn/pull/41789#discussion_r388636623) [](ancestors = 388636623)